Signal/src/ViewControllers/ConversationView/ConversationViewController.m

```diff
@@ -4171,11 +4171,19 @@ - (void)scrollToBottomAnimated:(BOOL)animated
 
     CGFloat contentHeight = self.safeContentHeight;
 
+    const CGFloat firstContentPageTop = ^{
+        if (@available(iOS 11, *)) {
+            return -self.collectionView.adjustedContentInset.top;
+        } else {
+            return -self.collectionView.contentInset.top;
+        }
+    }();
+
     // bottomLayoutGuide accounts for extra offset needed on iPhoneX
+    const CGFloat lastContentPageTop = contentHeight + self.collectionView.contentInset.bottom
+        + self.bottomLayoutGuide.length - self.collectionView.bounds.size.height;
 
-    CGFloat dstY = MAX(0,
```

Comment:
Our max used to be zero, but it needs to be relative to the content insets.

Reply:
Will the following all play together nicely:

* UIViewController manipulating the insets (UIViewController.automaticallyAdjustsScrollViewInsets) (is this only when the view is loaded?)
* Conversation view manipulating the insets in response to keyboard frame events.
* Conversation view manipulating the insets in response to "custom action sheet" events.

Reply:
Do we need to do a MIN() or MAX() to handle the edge cases where `contentHeight` is smaller than the viewport?

Reply:
Forgive me for thinking out loud, but I think what we're trying to do here is something like:

```
 const CGFloat viewportHeight = 
(self.collectionView.bounds.size.height - (self.collectionView.contentInset.bottom + self.bottomLayoutGuide.length));
const CGFloat lastContentPageTop = contentHeight - viewportHeight;
```
Is that clearer?

Reply:
> Forgive me for thinking out loud, but I think what we're trying to do here is something like:

Yup, that's what we're doing. Breaking out the temp variable as you have makes it clearer. I'll make that change.

Reply:
Also to be clear, this diff is bigger because I reformatted code, but the only functional difference is, instead of using 0 for the max offset, we use the self.collectionView.contentInset.top (which used to == 0)

Reply:
> Will the following all play together nicely:

In theory, insets are insets, so scroll to bottom should respect them just fine.

In practice I've run through these scenarios on iOS9, 10, 11x and everything looks good.